python/revng/scripting/projects/project.py

```diff
@@ -0,0 +1,151 @@
+#
+# This file is distributed under the MIT License. See LICENSE.md for details.
+#
+
+from abc import ABC, abstractmethod
+from copy import deepcopy
+from typing import Any, List
+
+from revng.model import Binary, get_element_by_path, make_diff
+from revng.tupletree import TypedList
+
+
+class Project(ABC):
+    def __init__(self) -> None:
+        self.model: Binary = Binary()
+        self.last_saved_model: Binary = deepcopy(Binary)
+        self.pipeline_description: dict = {}
+        self.input_binary: str = ""
+
+    @abstractmethod
+    def import_binary(self, input_binary: str):
+        pass
+
+    @abstractmethod
+    def get_artifact(self, artifact_name: str) -> bytes | str:
+        """
+        High level method that the user can call to fetch an artifact without targets.
+        """
+        pass
+
+    @abstractmethod
+    def _get_artifact(self, artifact_name: str, targets_class: List[Any] = []) -> bytes | str:
+        """
+        Resolves the target logic, performs MIME conversion and dispatches
+        the work to the `_get_artifact_internal` to actually get the artifact.
+        """
+        pass
+
+    @abstractmethod
+    def _get_artifact_internal(self, artifact_name: str, targets_class: List[Any] = []) -> bytes | str:
+        """
+        The actual implementation of how to fetch te artifacts, eg: call subprocess or gql.
+        """
+        pass
+
+    def get_artifacts(self, params: dict[str, List[Any]]) -> dict:
+        result = {}
+        for artifact_name, targets in params.items():
+            result[artifact_name] = self._get_artifact_internal(artifact_name, targets)
+
+        return result
+
+    @abstractmethod
+    def analyze(self, analysis_name: str, target_class: List[Any] = []) -> None:
+        pass
+
+    @abstractmethod
+    def analyses_list(self, analysis_name: str) -> None:
+        pass
+
+    @abstractmethod
+    def commit(self) -> None:
+        pass
+
+    def revert(self) -> None:
+        diff = make_diff(self.model, self.last_saved_model)
+        if not diff.changes:
+            return
+        self.set_model(deepcopy(self.last_saved_model))
+
+    @abstractmethod
+    def _get_pipeline_description(self) -> None:
+        pass
+
+    def import_and_analyze(self, input_binary_path: str):
+        self.import_binary(input_binary_path)
+        self.analyses_list("revng-initial-auto-analysis")
+        self.analyses_list("revng-c-initial-auto-analysis")
+        self._get_pipeline_description()
+        self.set_model_mixins()
+
+    def set_model(self, model: Binary) -> None:
+        self.model = model
+        self.last_saved_model = deepcopy(self.model)
+        # We call this method in multiple places, during the initial auto
+        # analysis for example we don't have the pipeline description yet
+        # so we need to take that into account.
+        if self.pipeline_description:
+            self.set_model_mixins()
+
+    def set_model_mixins(self) -> None:
```

Comment:
Is the user expected to call this? if not convert into a private method. Same for `set_model`.